Added environment but import failed

<!--- Provide a general summary of the issue in the Title above -->

## Description
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug. What were you hoping to achieve? -->

Adding an environment using the following JSON file, warns that some nodes couldn't be added, however it also allowed me to connect (when none of the nodes are visible in the Environments panel).

`nodes.json`:
```
[
    {
        "name": "peer0.org1.example.com",
        "api_url": "grpc://localhost:17052",
        "type": "fabric-peer",
        "msp_id": "Org1MSP",
    },
    {
        "name": "orderer.example.com",
        "api_url": "grpc://localhost:17050",
        "type": "fabric-orderer",
        "msp_id": "OrdererMSP",
    },
    {
        "name": "ca.org1.example.com",
        "api_url": "http://localhost:17054",
        "type": "fabric-ca",
        "ca_name": "ca.org1.example.com",
        "enroll_id": "admin",
        "enroll_secret": "adminpw"
    }
]
```


## Expected Behavior
<!--- Tell us what should happen -->
I shouldn't be able to connect if it wasn't able to read in the file correctly.

## Actual Behavior
<!--- Tell us what happens instead -->

```
[8/14/2019 10:25:46 AM] [ERROR] Error importing node file /Users/jake/Desktop/nodes.json: SyntaxError: /Users/jake/Desktop/nodes.json: Unexpected token } in JSON at position 159
[8/14/2019 10:25:46 AM] [WARNING] Added a new environment, but some nodes could not be added
[8/14/2019 10:25:54 AM] [INFO] connecting to fabric environment
[8/14/2019 10:25:54 AM] [SUCCESS] Connected to asdkasd
```

## Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

## Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- Reproduce this bug include code to reproduce, if relevant -->
1. Create a new environment using the `node.json` file
2. (error message)
3. Can seem to connect

## Existing issues
<!-- Have you searched for any existing issues or are their any similar issues that you've found? -->
- [ ] [Stack Overflow issues](https://stackoverflow.com/questions/tagged/ibp-vscode-extension)
- [ ] [GitHub Issues](https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues)

<!-- Please include any links to issues here -->

## Log File

<!-- In Visual Studio Code's command palette, run '>Developer: Open Logs Folder' -->
<!-- Attach/link to the latest nested 'Blockchain.log' file -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* IBM Blockchain Platform Extension Version:
* Visual Studio Code version:
* Environment name and version (e.g. Node.js v8.12.0, npm v6.4.1):
* Operating System and version:
* Link to your project:

--- a/README.md
+++ b/README.md
@@ -104,7 +104,7 @@ These some basic examples showing the required properties
     "name": "peer0.org1.example.com",
     "api_url": "grpc://localhost:17051",
     "type": "fabric-peer",
-    "msp_id": "Org1MSP",
+    "msp_id": "Org1MSP"
 }
 ```
 
@@ -113,7 +113,7 @@ These some basic examples showing the required properties
     "name": "orderer.example.com",
     "api_url": "grpc://localhost:17050",
     "type": "fabric-orderer",
-    "msp_id": "OrdererMSP",
+    "msp_id": "OrdererMSP"
 }
 ```
 
@@ -125,13 +125,13 @@ Here is how to have multiple nodes in one JSON file
         "name": "peer0.org1.example.com",
         "api_url": "grpc://localhost:17051",
         "type": "fabric-peer",
-        "msp_id": "Org1MSP",
+        "msp_id": "Org1MSP"
     },
     {
         "name": "orderer.example.com",
         "api_url": "grpc://localhost:17050",
         "type": "fabric-orderer",
-        "msp_id": "OrdererMSP",
+        "msp_id": "OrdererMSP"
     }
 ]
 

--- a/client/README.md
+++ b/client/README.md
@@ -100,7 +100,7 @@ These some basic examples showing the required properties
     "name": "peer0.org1.example.com",
     "api_url": "grpc://localhost:17051",
     "type": "fabric-peer",
-    "msp_id": "Org1MSP",
+    "msp_id": "Org1MSP"
 }
 ```
 
@@ -109,7 +109,7 @@ These some basic examples showing the required properties
     "name": "orderer.example.com",
     "api_url": "grpc://localhost:17050",
     "type": "fabric-orderer",
-    "msp_id": "OrdererMSP",
+    "msp_id": "OrdererMSP"
 }
 ```
 
@@ -121,13 +121,13 @@ Here is how to have multiple nodes in one JSON file
         "name": "peer0.org1.example.com",
         "api_url": "grpc://localhost:17051",
         "type": "fabric-peer",
-        "msp_id": "Org1MSP",
+        "msp_id": "Org1MSP"
     },
     {
         "name": "orderer.example.com",
         "api_url": "grpc://localhost:17050",
         "type": "fabric-orderer",
-        "msp_id": "OrdererMSP",
+        "msp_id": "OrdererMSP"
     }
 ]
 

--- a/client/src/commands/addEnvironmentCommand.ts
+++ b/client/src/commands/addEnvironmentCommand.ts
@@ -102,6 +102,15 @@ export async function addEnvironment(): Promise<{} | void> {
             }
         }
 
+        // check if any nodes were added
+        const newEnvironment: FabricEnvironment = new FabricEnvironment(environmentName);
+        const newNodes: FabricNode[] = await newEnvironment.getNodes();
+        if (newNodes.length === 0) {
+            await fs.remove(environmentPath);
+
+            throw new Error('no nodes were added');
+        }
+
         const fabricEnvironmentEntry: FabricEnvironmentRegistryEntry = new FabricEnvironmentRegistryEntry();
         fabricEnvironmentEntry.name = environmentName;
         await fabricEnvironmentRegistry.add(fabricEnvironmentEntry);

--- a/client/test/commands/addEnvironmentCommand.test.ts
+++ b/client/test/commands/addEnvironmentCommand.test.ts
@@ -37,11 +37,13 @@ describe('AddEnvironmentCommand', () => {
     let showInputBoxStub: sinon.SinonStub;
     let browseStub: sinon.SinonStub;
     let ensureDirStub: sinon.SinonStub;
+    let removeDirStub: sinon.SinonStub;
     let executeCommandSpy: sinon.SinonSpy;
     let sendTelemetryEventStub: sinon.SinonStub;
     let addMoreStub: sinon.SinonStub;
     let updateNodeStub: sinon.SinonStub;
     let readJsonStub: sinon.SinonStub;
+    let getNodesStub: sinon.SinonStub;
 
     before(async () => {
         await TestUtil.setupTests(mySandBox);
@@ -63,10 +65,22 @@ describe('AddEnvironmentCommand', () => {
             browseStub = mySandBox.stub(UserInputUtil, 'browse');
             addMoreStub = mySandBox.stub(UserInputUtil, 'addMoreNodes').resolves(UserInputUtil.DONE_ADDING_NODES);
             ensureDirStub = mySandBox.stub(fs, 'ensureDir').resolves();
+            removeDirStub = mySandBox.stub(fs, 'remove').resolves();
 
             executeCommandSpy = mySandBox.spy(vscode.commands, 'executeCommand');
             sendTelemetryEventStub = mySandBox.stub(Reporter.instance(), 'sendTelemetryEvent');
             updateNodeStub = mySandBox.stub(FabricEnvironment.prototype, 'updateNode').resolves();
+            getNodesStub = mySandBox.stub(FabricEnvironment.prototype, 'getNodes').resolves([{
+                short_name: 'peer0.org1.example.com',
+                name: 'peer0.org1.example.com',
+                api_url: 'grpc://localhost:17051',
+                chaincode_url: 'grpc://localhost:17052',
+                type: 'fabric-peer',
+                wallet: 'local_fabric_wallet',
+                identity: 'admin',
+                msp_id: 'Org1MSP',
+                container_name: 'fabricvscodelocalfabric_peer0.org1.example.com'
+            }]);
 
             readJsonStub = mySandBox.stub(fs, 'readJson').resolves({
                 short_name: 'peer0.org1.example.com',
@@ -102,6 +116,7 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_ENVIRONMENTS);
             ensureDirStub.should.have.been.calledOnce;
             updateNodeStub.should.have.been.calledOnce;
+            getNodesStub.should.have.been.calledOnce;
 
             logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
             logSpy.getCall(1).should.have.been.calledWith(LogType.SUCCESS, 'Successfully added a new environment');
@@ -149,6 +164,7 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_ENVIRONMENTS);
             ensureDirStub.should.have.been.calledOnce;
             updateNodeStub.should.have.been.calledTwice;
+            getNodesStub.should.have.been.calledOnce;
 
             logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
             logSpy.getCall(1).should.have.been.calledWith(LogType.SUCCESS, 'Successfully added a new environment');
@@ -183,6 +199,7 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_GATEWAYS);
             ensureDirStub.should.have.been.calledTwice;
             updateNodeStub.should.have.been.calledTwice;
+            getNodesStub.should.have.been.calledTwice;
             logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
             logSpy.getCall(1).should.have.been.calledWith(LogType.SUCCESS, 'Successfully added a new environment');
             logSpy.getCall(2).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
@@ -214,6 +231,7 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_ENVIRONMENTS);
             ensureDirStub.should.have.been.calledOnce;
             updateNodeStub.should.have.been.calledTwice;
+            getNodesStub.should.have.been.calledOnce;
 
             addMoreStub.should.have.been.calledTwice;
 
@@ -336,6 +354,7 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_ENVIRONMENTS);
             ensureDirStub.should.have.been.calledOnce;
             updateNodeStub.should.have.been.calledOnce;
+            getNodesStub.should.have.been.calledOnce;
 
             logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
             logSpy.getCall(1).should.have.been.calledWith(LogType.SUCCESS, 'Successfully added a new environment');
@@ -383,11 +402,48 @@ describe('AddEnvironmentCommand', () => {
             executeCommandSpy.should.have.been.calledWith(ExtensionCommands.REFRESH_ENVIRONMENTS);
             ensureDirStub.should.have.been.calledOnce;
             updateNodeStub.should.have.been.calledOnce;
+            getNodesStub.should.have.been.calledOnce;
 
             logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
-            logSpy.should.have.been.calledWith(LogType.ERROR,  `Error importing node file ${uri.fsPath}: A node should have a name property`, `Error importing node file ${uri.fsPath}: Error: A node should have a name property`);
+            logSpy.should.have.been.calledWith(LogType.ERROR, `Error importing node file ${uri.fsPath}: A node should have a name property`, `Error importing node file ${uri.fsPath}: Error: A node should have a name property`);
             logSpy.should.have.been.calledWith(LogType.WARNING, 'Added a new environment, but some nodes could not be added');
             sendTelemetryEventStub.should.have.been.calledOnceWithExactly('addEnvironmentCommand');
+        });
+
+        it('should not add an environment if cannot add any nodes', async () => {
+            getNodesStub.resolves([]);
+            showInputBoxStub.onFirstCall().resolves('myEnvironment');
+
+            readJsonStub.resolves({
+                short_name: 'invalid',
+                api_url: 'grpc://localhost:17051',
+                chaincode_url: 'grpc://localhost:17052',
+                type: 'fabric-peer',
+                wallet: 'local_fabric_wallet',
+                identity: 'admin',
+                msp_id: 'Org1MSP',
+                container_name: 'fabricvscodelocalfabric_peer0.org1.example.com'
+            });
+
+            const uri: vscode.Uri = vscode.Uri.file(path.join('myPath'));
+            browseStub.onFirstCall().resolves([uri]);
+
+            await vscode.commands.executeCommand(ExtensionCommands.ADD_ENVIRONMENT);
+
+            const environments: Array<any> = vscode.workspace.getConfiguration().get(SettingConfigurations.FABRIC_ENVIRONMENTS);
+
+            environments.length.should.equal(0);
+
+            ensureDirStub.should.have.been.calledOnce;
+            updateNodeStub.should.not.have.been.called;
+            removeDirStub.should.have.been.called;
+
+            const error: Error = new Error('no nodes were added');
+
+            logSpy.getCall(0).should.have.been.calledWith(LogType.INFO, undefined, 'Add environment');
+            logSpy.should.have.been.calledWith(LogType.ERROR, `Error importing node file ${uri.fsPath}: A node should have a name property`, `Error importing node file ${uri.fsPath}: Error: A node should have a name property`);
+            logSpy.should.have.been.calledWith(LogType.ERROR, `Failed to add a new environment: ${error.message}`, `Failed to add a new environment: ${error.toString()}`);
+            sendTelemetryEventStub.should.not.have.been.called;
         });
     });
 });

